Fix bizarre custom anamorphic behavior

jstebbins opened the pull request:
When "keep aspect" is unset in custom anamorphic, it was making very
uintuitive changes to PAR.  This simplifies the code and makes the
behavior more sane.

jstebbins committed:
Fix bizarre custom anamorphic behavior

When "keep aspect" is unset in custom anamorphic, it was making very
uintuitive changes to PAR.  This simplifies the code and makes the
behavior more sane.

bradleysepos commented:
Did you check blame to see why you added/updated most of this previously?

https://github.com/HandBrake/HandBrake/blame/master/libhb/hb.c

jstebbins commented:
None of the commit messages are particularly helpful.  But I know that I've never done very extensive testing of custom-ana with keep-aspect off.  It's a use case that has few legitimate uses.

bradleysepos commented:
I've used custom ana a number of times to fix a source with improperly set anamorphic values. A number of DVD masters have this problem. As far as I can remember, that is the most common use case and why custom anamorphic was implemented.

On one occasion I used custom ana to store a 1920x1080 video as 1440x1080 pixels due to storage constraints. Preserved more fine detail and grain than increasing crf.

Anyway, I'm willing to test a variety of scenarios when I have a moment.

bradleysepos commented:
One issue I found in testing ghb is that when reducing display width, the preview height appears to grow.

jstebbins committed:
LinGui: custom-ana reduce PAR when it's modified indirectly

If PAR is indirectly changed through the user changing storage or
display width and height, reduce PAR to minimum num and den.

jstebbins committed:
libhb: prevent crashes in hb_get_preview2

Try to avoid failures to initialize sws context by setting minimum
dimensions.  And if initialization does fail, exit gracefully instead of
crashing in sws_scale.

jstebbins committed:
LinGui: don't call hb_get_preview2 with outragious dimensions

bradleysepos commented:
Ship it!

jstebbins committed:
Enforce min/max dimensions

In both hb_set_anamorphic_size and hb_get_preview, check that image
dimensions are > 32x32 and < 10240x10240.  This should help to prevent
crashes with unusual dimensions settings.

jstebbins committed:
Double HB_MAX_WIDTH/HEIGHT

Per BradleyS suggestion

bradleysepos commented:
Ship it!

For posterity, I guess here is as good a place as anywhere.

20480x20480 (419.4 megapixels) is double the width of 21:9 ratio 8k (10240×4320, 44.2 megapixels) and is considerably larger than even 1:1 ratio fulldome 8k (8192x8192, 67.1 megapixels). The only current application for supersize video frames aside from top tier professional camera output is timelapse video from ultra-high resolution still images—one project I found boasts 10328x7760 source resolution using a PhaseOne IQ180. Suffice to say, if we ever need to process larger frames than 20k*20k, it's definitely The Future™.

I upscaled an 848x480 source to 20480x11592 (237.4 megapixels) using latest CLI on OS X. It took 8 minutes to encode 150 frames with reasonable x264 settings on a 6x3.33 GHz CPU. Aside from a few encoder warnings, everything worked. MacCLI and MacGUI both re-ingested the result and happily scaled back down to the original resolution with no problems, despite taking 14 seconds to generate each image preview during scan. The preview window worked but was slightly less responsive, scaling the huge source down to 7% size to fit on my 1080p display. Pretty cool, albeit insane.
